### ISSUE TITLE
Caching shouldn't be used when a task has retires and in-non completed state

### ISSUE THREAD

**Issue by BitTheByte** [PrefectHQ/prefect#9012]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

This issue applies when using `prefect_dask` I'm not sure if it's occurring somewhere else when a task is submitted in a dask cluster and has retires configured if that task fails for any reason e.g. dask just killed the task or even the worker is killed if it has the time to change state to "Crashed" a cache entry of this task will be created and when its resubmitted for execution again e.g. flow restarts it will use the cached "Crashed" state even if it still has some retires.

### Reproduction

```python3
-
```

### Versions

```Text
Version:             2.9.0
API version:         0.8.4
Python version:      3.11.2
Git commit:          69f57bd5
Built:               Thu, Mar 30, 2023 1:08 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         cloud
```

**Comment by BitTheByte** [comment]
https://github.com/PrefectHQ/prefect/issues/8821 is affected by this issue

**Comment by madkinsz** [comment]
Hey @BitTheByte I think this is probably Dask caching the result of the task run not us.

Can you please provide MREs with your issues? We cannot prioritize resolving issues without minimal examples.

**Comment by BitTheByte** [comment]
I think it's prefect as I could see
```
15:13:23.080 | INFO    | Task run 'dummy_task-0' - Task run '6ff789a3-7f56-42e7-aed6-b918414487c3' already finished.
```
being logged to the could, I believe this means dask has executed the future and prefect decided to use its cache. I'm still working on the minimal example, this might take a while as the behavior is unpredictable

**Comment by madkinsz** [comment]
@BitTheByte Ah I see what you mean — this is not a "cache" persay, it's just the state of the run itself. 

@billpalombi / @rpeden perhaps similar to our recent changes around restart transitions we should attempt transitions from CRASHED -> RUNNING if the client finds itself being asked to run a crashed task run.

**Comment by billpalombi** [comment]
Yes, I agree. That's consistent with the choice we made in #8802.

Is it just a matter or removing the rule, or are there other implementation considerations? If its just the rule, this could be a quick win.

**Comment by BitTheByte** [comment]
@billpalombi 

I believe this would also fix https://github.com/PrefectHQ/prefect/issues/9014, I've tested it locally, but keep in mind allowing tasks to transition from "Crashed" to "Running" miss up the internal state on both dask and prefect, I think this might need some engineering first

**Comment by BitTheByte** [comment]
@billpalombi 

I believe this would also fix https://github.com/PrefectHQ/prefect/issues/9014, I've tested it locally.

**Comment by BitTheByte** [comment]
@madkinsz Can we get this fixed before the 2.10 release? Its fix should be trivial

**Comment by madkinsz** [comment]
@BitTheByte sorry but we do not make last minute changes on planned release days — we'll definitely look into this for the next release and we can always release early.
